library/core/src/option.rs

```diff
@@ -1869,6 +1869,121 @@ impl<T, E> Option<Result<T, E>> {
     }
 }
 
+/// Implements the various `Fn` traits to make the `transpose_fn` family of
+/// methods work over any functions.
+///
+/// This is not otherwise possible with a closure due to `|args: A|` being
+/// treated as `Fn<(A,)>` instead of `Fn<A>`.
+struct TransposeFn<F> {
+    f: Option<F>,
+}
+
+impl<F> Option<F> {
+    /// Transposes an `Option` of an [`Fn`] into an [`Fn`] that returns an
+    /// `Option`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(transpose_fn)]
+    ///
+    /// let x = Some(|| 42);
+    /// let y = x.transpose_fn();
+    ///
+    /// assert_eq!(y(), Some(42));
+    /// ```
+    #[inline]
+    #[unstable(feature = "transpose_fn", issue = "none")]
+    pub fn transpose_fn<A>(self) -> impl Fn<A, Output = Option<F::Output>>
+    where
+        F: Fn<A>,
+        A: Tuple,
+    {
+        impl<A, F> Fn<A> for TransposeFn<F>
+        where
+            F: Fn<A>,
+            A: Tuple,
+        {
+            #[inline]
+            extern "rust-call" fn call(&self, args: A) -> Self::Output {
+                self.f.as_ref().map(|f| f.call(args))
+            }
```

Comment:
Note: `f(args)` is not allowed here because that would be for calling `Fn<(A,)>`